### DISCUSSION TITLE
Custom Image Resolver

### DISCUSSION THREAD

**Discussion by tekguy** [SixLabors/ImageSharp.Web#132] [Q&A]
Hello,

I recently upgraded to version 1.0.1 of the ImageSharp.Web nuget. I  have two custom Image Providers  (ProviderA, ProviderB). 
Both ORIGINAL images I am grabbing updates the image every 5 seconds or so (think security cams).

I am trying to figure out a way to set the max cache for each provider (one for 10 seconds, one for 1 min as an example). Here is the code I am using in ProviderA:

```
public async Task<ImageMetadata> GetMetaDataAsync()
{
    // just randomly setting a content length to get it to update the disk cache
    var rand = new Random(1000);
    var contentLength = rand.Next();


    var maxAge = TimeSpan.FromTicks(DateTime.UtcNow.AddSeconds(10).Ticks);
    var meta = new ImageMetadata(DateTime.UtcNow, maxAge, contentLength);

    return await Task.FromResult(meta);
}
```
I keep getting the following in my debug output when refreshing after 10 seconds:
SixLabors.ImageSharp.Web.Middleware.ImageSharpMiddleware: Information: The image 'https://BLAH/BLAH.jpg?format=jpg&width=360' was not modified

With this code, I would expect the image to update every 10 seconds, but it is not. The image only gets updated when I restart the app. 

I was able to get the image to update every 10 seconds by applying:

**options.CacheMaxAge = TimeSpan.FromSeconds(10);**

Here is my ConfigureServices

```

public void ConfigureServices(IServiceCollection services)
{
    services.AddImageSharp(options => {
        options.CacheMaxAge = TimeSpan.FromSeconds(10);
    })
    .ClearProviders()
    .Configure<AzureBlobStorageImageProviderOptions>(options =>
        {
            options.BlobContainers.Add(new AzureBlobContainerClientOptions
            {
                ConnectionString = "OMIT",
                ContainerName = "OMIT"
            });
        })
    .Configure<PhysicalFileSystemCacheOptions>(options =>
    {
        options.CacheFolder = "is-cache";
    })
    .SetCache(provider =>
    {
        return new PhysicalFileSystemCache(
            provider.GetRequiredService<IOptions<PhysicalFileSystemCacheOptions>>(),
            provider.GetRequiredService<IWebHostEnvironment>(),
            provider.GetRequiredService<IOptions<ImageSharpMiddlewareOptions>>(),
            provider.GetRequiredService<FormatUtilities>());
    })
    .SetCacheHash<CacheHash>()
    .AddProvider<AzureBlobStorageImageProvider>()
    .AddProvider<ProviderA>()
    .AddProvider<ProviderB>()
    .AddProcessor<ResizeWebProcessor>()
    .AddProcessor<FormatWebProcessor>()
    .AddProcessor<BackgroundColorWebProcessor>();
}

```

So with that said. Can I somehow apply the CacheMaxAge per provider in the ConfigureService? Or am I doing something wrong within the GetMetaDataAsync method?

Thank you!

**Comment by JimBobSquarePants** [comment]
Ok...

So yeah, you cannot currently set the cached max age per image or provider only the max age that is sent to the browser headers collection.

However, I think it's reasonable to request that the max age override affects both browser and cache values. 

We could add a property to the options that chooses whether we look at the global `CacheMaxAge` property or the CacheControlMaxAge property of the cache metadata (_which is set by the source provider_) when we check to see whether the cache has expired.  

Then you could set individual providers to cache images for individual times both in the browser and cache.
